From ef0e1ad4c687338b2a86b03882e1355dd7667f9d Mon Sep 17 00:00:00 2001
From: Mike Yoder <myoder@cloudera.com>
Date: Tue, 3 Nov 2015 14:35:25 -0800
Subject: [PATCH 0948/1023] CLOUDERA-BUILD. CDH-33705. HttpFS Keystore
 password exposed on cmd line

The keystore password for httpfs was placed into $CATALINA_OPTS, which
ultimately makes its way to the command line as Tomcat starts up. This
changes things so that if CATALINA_BASE (the environment variable
describing where Catalina config information is located) exists, and
the catalina.properties file exists, then the appropriate config
entries are written into the catalina.properties file. Information
placed in that file goes into Tomcat just like the environment
variable does.

Change-Id: I7dfdc99fe6fbdbbec4ce04a6d46ad97d52eb6381
(cherry picked from commit 77d25b45d3b6abea8fd0f7a9ba64bc32162c61f4)
---
 .../src/main/libexec/httpfs-config.sh              |    2 +-
 .../hadoop-hdfs-httpfs/src/main/sbin/httpfs.sh     |   56 ++++++++++++++++++--
 2 files changed, 54 insertions(+), 4 deletions(-)

diff --git a/hadoop-hdfs-project/hadoop-hdfs-httpfs/src/main/libexec/httpfs-config.sh b/hadoop-hdfs-project/hadoop-hdfs-httpfs/src/main/libexec/httpfs-config.sh
index a2fe1c2..94a97b8 100644
--- a/hadoop-hdfs-project/hadoop-hdfs-httpfs/src/main/libexec/httpfs-config.sh
+++ b/hadoop-hdfs-project/hadoop-hdfs-httpfs/src/main/libexec/httpfs-config.sh
@@ -161,7 +161,7 @@ if [ "${HTTPFS_SSL_KEYSTORE_PASS}" = "" ]; then
   export HTTPFS_SSL_KEYSTORE_PASS=password
   print "Setting HTTPFS_SSL_KEYSTORE_PASS:     ${HTTPFS_SSL_KEYSTORE_PASS}"
 else
-  print "Using   HTTPFS_SSL_KEYSTORE_PASS:     ${HTTPFS_SSL_KEYSTORE_PASS}"
+  print "Using   HTTPFS_SSL_KEYSTORE_PASS:     *REDACTED*"
 fi
 
 if [ "${CATALINA_BASE}" = "" ]; then
diff --git a/hadoop-hdfs-project/hadoop-hdfs-httpfs/src/main/sbin/httpfs.sh b/hadoop-hdfs-project/hadoop-hdfs-httpfs/src/main/sbin/httpfs.sh
index a593b67..ef0b7fc 100644
--- a/hadoop-hdfs-project/hadoop-hdfs-httpfs/src/main/sbin/httpfs.sh
+++ b/hadoop-hdfs-project/hadoop-hdfs-httpfs/src/main/sbin/httpfs.sh
@@ -13,6 +13,54 @@
 #  limitations under the License.
 #
 
+# Attempt to set a value in the catalina.properties file. Insert new
+# attributes; overwrite existing attributes.
+# $1 attribute
+# The rest - value
+# return: 1 on failure, 0 on success
+set_catalina_prop() {
+  ATTR=$1
+  shift
+  VALUE="$@"
+  if [ ! -e "${CATALINA_BASE}" ]; then
+    return 1
+  fi
+  CATALINA_PROPERTIES="${CATALINA_BASE}/conf/catalina.properties"
+  if [ ! -e "${CATALINA_PROPERTIES}" ]; then
+    return 1
+  fi
+  if grep -q "^${ATTR}=" "${CATALINA_PROPERTIES}"; then
+    sed -i "s#${ATTR}=.*#${ATTR}=${VALUE}#" "${CATALINA_PROPERTIES}"
+    if [ $? -eq 1 ]; then
+      # sed somehow fails on the replace; fail so environment is used.
+      return 1
+    fi
+  else
+    echo "${ATTR}=${VALUE}" >> "${CATALINA_PROPERTIES}"
+  fi
+  return 0
+}
+
+# Try to set a value in the catalina.properties file. If that fails, set
+# it in catalina_opts: an environment variable that becomes a command
+# line argument.
+# $1 attribute
+# The rest - the value
+tomcat_set_prop() {
+  ATTR=$1
+  shift
+  VALUE="$@"
+  # If no value, don't set anything.
+  if [ -z "${VALUE}" ]; then
+    return
+  fi
+
+  set_catalina_prop "${ATTR}" "${VALUE}"
+  if [ $? -eq 1 ]; then
+    catalina_opts="${catalina_opts} -D${ATTR}=${VALUE}";
+  fi
+}
+
 # resolve links - $0 may be a softlink
 PRG="${0}"
 
@@ -43,9 +91,11 @@ catalina_opts="${catalina_opts} -Dhttpfs.temp.dir=${HTTPFS_TEMP}";
 catalina_opts="${catalina_opts} -Dhttpfs.admin.port=${HTTPFS_ADMIN_PORT}";
 catalina_opts="${catalina_opts} -Dhttpfs.http.port=${HTTPFS_HTTP_PORT}";
 catalina_opts="${catalina_opts} -Dhttpfs.http.hostname=${HTTPFS_HTTP_HOSTNAME}";
-catalina_opts="${catalina_opts} -Dhttpfs.ssl.enabled=${HTTPFS_SSL_ENABLED}";
-catalina_opts="${catalina_opts} -Dhttpfs.ssl.keystore.file=${HTTPFS_SSL_KEYSTORE_FILE}";
-catalina_opts="${catalina_opts} -Dhttpfs.ssl.keystore.pass=${HTTPFS_SSL_KEYSTORE_PASS}";
+
+# Try to put SSL items inside catalina.properties; on failure fall back to command line.
+tomcat_set_prop httpfs.ssl.enabled "${HTTPFS_SSL_ENABLED}"
+tomcat_set_prop httpfs.ssl.keystore.file "${HTTPFS_SSL_KEYSTORE_FILE}"
+tomcat_set_prop httpfs.ssl.keystore.pass "${HTTPFS_SSL_KEYSTORE_PASS}"
 
 print "Adding to CATALINA_OPTS:     ${catalina_opts}"
 
-- 
1.7.9.5

